p4-16/spec/P4-16-spec.mdk

```diff
@@ -4946,21 +4946,69 @@ switchCase
 
 switchLabel
     : name
+    | expression
     | DEFAULT
     ;
 ~ End P4Grammar
 
-The expression within the `switch` statement is restricted to be
-the result of a table invocation (See Section [#sec-invoke-mau]).
-
-If a switch label is not followed by a block statement it falls
+The expression within the `switch` statement must be one of:
+
+- an expression of the form `t.apply().action_run`, where `t` is the name
+  of a table (see Section [#sec-invoke-mau])
+- an expression that evaluates to a result with one of the types
+  listed below:
+  - `bit<W>`
+  - `int<W>`
+  - `enum`, either with or without an underlying representation specified
+  - `error`
+
+If the `switch` expression is of the form `t.apply().action_run`, then
+all switch labels must be names of actions of the table `t`.
+Otherwise, all switch labels must be expressions with compile-time
+known values, and have the same type as the type of the switch
+expression.
+
+TBD: What language to use above if we want to allow a switch
```

Comment:
you could say that they must have a type which can be implicitly cast to the switch expression type.

Reply:
Thanks for that.  Incorporated that terminology into the latest commit.